### ISSUE TITLE
os: ExpandEnv: Missing curly bracket at end of string substitutes var name whether env var is set or not.

### ISSUE THREAD

**Issue by gfl-chris** [golang/go#35752]
### What version of Go are you using (`go version`)?

<pre>
go version go1.12.9 linux/amd64
</pre>
and
<pre>
Go Playground (1.13.4)
</pre>

### Does this issue reproduce with the latest release?
Yes. (Go Playground)

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/ceineke/.cache/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/ceineke/.gvm/pkgsets/go1.12.9/global"
GOPROXY=""
GORACE=""
GOROOT="/home/ceineke/.gvm/gos/go1.12.9"
GOTMPDIR=""
GOTOOLDIR="/home/ceineke/.gvm/gos/go1.12.9/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build732428512=/tmp/go-build -gno-record-gcc-switches"
</pre>
and
<pre>
Go Playground
</pre></details>

### What did you do?
https://play.golang.org/p/GSfBKiZqj2L

### What did you expect to see?
Either
```
Hello, ${name
Hello, ${name
```
or
```
Hello, ${name
Hello, foo
```

### What did you see instead?
```
Hello, name
Hello, name
```

**Comment by cuonglm** [comment]
`{` is an undefined environment variable, so `${` is expanded to empty string. I think it's working as intended. Environment variable is not the same as shell variable.

**Comment by gfl-chris** [comment]
It's not and it shouldn't. The code [here](https://golang.org/src/os/env.go#L87) points out that `${` is bad syntax. I guess what I'm trying to say is that it shouldn't be "eaten."

Concrete example: Running ExpandEnv over a YAML file whoselast line looks like this:
```
  password: abcd${efgh
```
will, IMHO, clobber the password.

**Comment by cuonglm** [comment]
> It's not and it shouldn't. The code [here](https://golang.org/src/os/env.go#L87) points out that `${` is bad syntax. I guess what I'm trying to say is that it shouldn't be "eaten."
> 
> Concrete example: Running ExpandEnv over a YAML file whose last line looks like this:
> 
> ```
>   password: abcd${efgh
> ```
> 
> will, IMHO, incorrectly clobber the password.

Hmm, it's also not consistent, since when we can do `os.Setenv("{", "foo")` then `os.Getenv("{")` will return `foo`, but `ExpandEnv` does not.

I think it's better to just improve the document of `[ExpandEnv](https://golang.org/pkg/os/#ExpandEnv)` to indicate that the expansion follow the rule of shell expansion, invalid syntax will return empty string.

**Comment by ianlancetaylor** [comment]
I agree that we should mention this in the documentation and not change the behavior of the function.

**Comment by opennota** [comment]
In another case Expand was fixed:
https://github.com/golang/go/issues/24345
Perhaps in this case ExpandEnv should be fixed too.

**Comment by ianlancetaylor** [comment]
There is no significant difference between `Expand` and `ExpandEnv`.

See also #26135.

That said, I think you are suggesting that for `${` with no `}` we should keep the `${`.  That may be OK.

**Comment by opennota** [comment]
> you are suggesting that for `${` with no `}` we should keep the `${`.

Yep. It's not even like ExpandEnv replaces `${` with a value of the `{` variable: https://play.golang.org/p/CzLSGzkLAWx

**Comment by fitzy101** [comment]
I'm looking at picking up an issue for a first contribution, and it looks as though this is a candidate. Is that actively being addressed by anyone? It looks like there is still a decision to be made on the expected behaviour.
